### PR TITLE
HPCC-14833 Reduce reallocation of hashtables when freeing

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith please review.

File: system/jlib/jsuperhash.cpp [hunk 1]
```diff
@@ -483,26 +483,36 @@ void SuperHashTable::doKill(void)
...
 void SuperHashTable::kill(void)
 {
     releaseAll();
-    doKill();
-    tablesize = InitialTableSize;
-    tablecount = 0;
-    setCache(0);
-    table = (void * *) checked_malloc(InitialTableSize*sizeof(void *),-604);
-    memset(table,0,InitialTableSize*sizeof(void *));
+    if (tablesize != InitialTableSize)
```
Comment (ghalliday) [inline system/jlib/jsuperhash.cpp]: avoid kill() reallocating a table that is already the correct size (e.g., if it has already been killed)

File: system/jlib/jsuperhash.cpp [hunk 2]
```diff
@@ -483,26 +483,36 @@ void SuperHashTable::doKill(void)
 
 void SuperHashTable::releaseAll(void)
 {
-    unsigned i;
-    for (i = 0; i < tablesize; i++)
+    if (tablecount)
```
Comment (ghalliday) [inline system/jlib/jsuperhash.cpp]: avoid walking the table if there are no elements (e.g., if already cleared)

**Comment by richardkchapman** [comment]
Why is there a releaseElements() and a releaseAll() ?

**Comment by ghalliday** [comment]
I started off with just releaseAll().  However, the thread-safe versions of the hash tables should wrap that function with a critical section, but there is no need to protect a call in a destructor.

So releaseAll() is protected, and can be called from the class destructor.  releaseElements() is public and can be called from anywhere.  I'm no so keen on the name, but I didn't come up with anything better.

**Comment by richardkchapman** [comment]
Perhaps you should make that clear in comments?

I would probably have gone for releaseAll (public) and _releaseAll (private)

**Comment by richardkchapman** [comment]
@ghalliday Back to you for now

File: system/jlib/jsuperhash.cpp [hunk 3]
```diff
@@ -483,26 +483,36 @@ void SuperHashTable::doKill(void)
...
+        unsigned i;
+        for (i = 0; i < tablesize; i++)
+        {
+            void * et = table[i];
+            table[i] = NULL;
+            if (et)
+                onRemove(et);
+        }
+        tablecount = 0;
+        setCache(0);
```
Comment (jakesmith) [inline system/jlib/jsuperhash.cpp]: Not resetting the cache (previous behaviour), meant a potential inconsequential condition in doFindExact being met.
Is that the only consequence?

Trivial, but was slightly faster before when left set (>tablesize) as it meant doFindExact avoided the 2nd expression test (table[i] != et)

**Comment by jakesmith** [comment]
>~SimpleHashTableOf<ET, FP>() { SELF::kill (); }

still calling kill(), should call releaseAll() shouldn't it?

**Comment by jakesmith** [comment]
> ~SimpleHashTableOf() { SELF::kill (); }

> ~OwningSimpleHashTableOf<ET, FP>() { SELF::kill(); }

>~StringSuperHashTableOf<ET>() { SELF::kill(); }

>    ~OwningStringSuperHashTableOf<ET>() { SELF::kill(); }

should all call releaseAll() I think.

**Comment by jakesmith** [comment]
@ghalliday - looks good, but I think a few more kill()'s could be removed.

**Comment by jakesmith** [comment]
>I would probably have gone for releaseAll (public) and _releaseAll (private)

I would likely have chosen that naming also.

**Comment by ghalliday** [comment]
@jakesmith in sasha/packetstore.cpp CDataPacketStore's destructor doesn't call _releaseAll().  I assume it should, but is it safe to include in this PR?

@richardkchapman first commit is the same with the functions renamed.  Second commit adds some extra calls suggested by Jake.

**Comment by jakesmith** [comment]
>@jakesmith in sasha/packetstore.cpp CDataPacketStore's destructor doesn't call _releaseAll(). I assume it should, but is it safe to include in this PR?


Very safe.. It's not used. Not it's owning class - I've don't know if ever been used..
May be best to delete packetstore.* altogether.

**Comment by jakesmith** [comment]
There may be a good reason I forget, but requiring releaseAll in derived destructor, looks like a good reason it should be called in beforeDispose() in base class.

..except I guess derived destructor *might* do something with elems in dtor..

**Comment by ghalliday** [comment]
Yes, if there is a beforeDispose() function then it would be sensible to call it from there.  The new checks mean extra calls will have little performance impact.

Really the design is wrong.  It should either use a destructor on the contained object to achieve everything.  (E.g., a hash table of owned elements), or possibly it could use a function in an associated traits class to perform the removal/addition.  At some point it may be worth biting the bullet and refactor the implementation.  (Arrays could do with some similar treatment.)

**Pull request merged**